impeller/docs/babys_first_triangle.md

```diff
@@ -0,0 +1,276 @@
+
+Notice that all these interfaces and metadata are in a struct called `BabyVertexShader`. Find a similar struct called `BabyFragmentShader` in `baby.frag.h`. [Tinker around with the shader in the compiler explorer](https://tinyurl.com/28fypq2b) to see what the compiler generates.
+
+Now, let's put together a pipeline. First you need a pipeline descriptor.
+
+```c++
+// Declare a shorthand for the shaders we are going to use.
+using VS = BabyVertexShader;
+using FS = BabyFragmentShader;
+
+// Create a pipeline descriptor that uses the shaders together and default
+// initializes the fixed function state.
+//
+// If the vertex shader outputs disagree with the fragment shader inputs, this
+// will be a compile time error.
+auto desc = PipelineBuilder<VS, FS>::MakeDefaultPipelineDescriptor(*context);
+```
+
+`MakeDefaultPipelineDescriptor` fills in the default values for the fixed function stages for the pipeline. Depending on the environment, some minor tweaks to this descriptor might be necessary.
+
+Astute readers will notice that this is the first time we have tried to fuse the vertex and fragment stages together. But what happens if they aren't compatible? For instance, what if the fragment stage expects an input the vertex stage doesn't provide? The compiler has thought of this for you and the act of putting together a pipeline builder via `PipelineBuilder<VS, FS>` will statically (at build time) perform the check for you. If the stages aren't compatible the C++ code will refuse to compile.
+
+Finally, we have everything we need, let's create the pipeline:
+
+```c++
+auto pipeline = context->GetPipelineLibrary()->GetPipeline(desc).Get();
+```
+
+This call creates a pipeline on a background thread and returns a future. But we need the pipeline right away. Just `Get` wait on the pipeline creation. Remember, pipelines are expensive to create. So just do this once upfront and keep them around for as long as possible. And don't create these in frame workload unless you want jank.
+
+### Vertex Data
+
+Remember the `PerVertexData` struct the compiler generated for us? It is our job to fill in the the vertex information and provide it as a GPU buffer allocation to the draw call. Handily, there is another C++ utility called the `VertexBufferBuilder<T>` that can create this buffer for us. The template parameter T is the `PerVertexData` struct.
+
+Let's create the data we will be giving our draw call. Remember, our vertex shader is a bit of slacker and doesn't do anything to the vertex information to convert it to normalized device coordinates. So we need to make sure the information is already in [normalized device coordinate](coordinate_system.md).
+
+```c++
+VertexBufferBuilder<VS::PerVertexData> vertex_buffer_builder;
```

Comment:
interestingly this class is definitely on my list of abstractions to remove. I was thinking about some kind of either template or list initalization that would avoid the secondary std::vec allocation.

No action needed, just a note for me to update the doc if I do that.